File: generated-code/codecraft/typescript/TcpReadWrite/src/model/entity-properties.ts
import { AttackProperties } from "./attack-properties";
import { BuildProperties } from "./build-properties";
import { RepairProperties } from "./repair-properties";
import { Stream } from "../stream";

/**
 * Entity properties
 */
export class EntityProperties {
    /**
     * Size. Entity has a form of a square with side of this length
     */
    size: number
    /**
     * Score for building this entity
     */
    buildScore: number
    /**
     * Score for destroying this entity
     */
    destroyScore: number
    /**
     * Whether this entity can move
     */
    canMove: boolean
    /**
     * Number of population points this entity provides, if active
     */
    populationProvide: number
    /**
     * Number of population points this entity uses
     */
    populationUse: number
    /**
     * Maximum health points
     */
    maxHealth: number
    /**
     * Cost to build this first entity of this type. If this is a unit (entity can move), the cost is increased by 1 for each existing unit of this type
     */
    initialCost: number
    /**
     * If fog of war is enabled, maximum distance at which other entities are considered visible
     */
    sightRange: number
    /**
     * Amount of resource added to enemy able to collect resource on dealing damage for 1 health point
     */
    resourcePerHealth: number
    /**
     * Build properties, if entity can build
     */
    build: BuildProperties | null
    /**
     * Attack properties, if entity can attack
     */
    attack: AttackProperties | null
    /**
     * Repair properties, if entity can repair
     */
    repair: RepairProperties | null

    constructor(size: number, buildScore: number, destroyScore: number, canMove: boolean, populationProvide: number, populationUse: number, maxHealth: number, initialCost: number, sightRange: number, resourcePerHealth: number, build: BuildProperties | null, attack: AttackProperties | null, repair: RepairProperties | null) {
        this.size = size;
        this.buildScore = buildScore;
        this.destroyScore = destroyScore;
        this.canMove = canMove;
        this.populationProvide = populationProvide;
        this.populationUse = populationUse;
        this.maxHealth = maxHealth;
        this.initialCost = initialCost;
        this.sightRange = sightRange;
        this.resourcePerHealth = resourcePerHealth;
        this.build = build;
        this.attack = attack;
        this.repair = repair;
    }

    /**
     * Read EntityProperties from input stream
     */
    static async readFrom(stream: Stream): Promise<EntityProperties> {
        let size;
        size = await stream.readInt();
        let buildScore;
        buildScore = await stream.readInt();
        let destroyScore;
        destroyScore = await stream.readInt();
        let canMove;
        canMove = await stream.readBool();
        let populationProvide;
        populationProvide = await stream.readInt();
        let populationUse;
        populationUse = await stream.readInt();
        let maxHealth;
        maxHealth = await stream.readInt();
        let initialCost;
        initialCost = await stream.readInt();
        let sightRange;
        sightRange = await stream.readInt();
        let resourcePerHealth;
        resourcePerHealth = await stream.readInt();
        let build;
        if (await stream.readBool()) {
            build = await BuildProperties.readFrom(stream);
        } else {
            build = null;
        }
        let attack;
        if (await stream.readBool()) {
            attack = await AttackProperties.readFrom(stream);
        } else {
            attack = null;
        }
        let repair;
        if (await stream.readBool()) {
            repair = await RepairProperties.readFrom(stream);
        } else {
            repair = null;
        }
        return new EntityProperties(size, buildScore, destroyScore, canMove, populationProvide, populationUse, maxHealth, initialCost, sightRange, resourcePerHealth, build, attack, repair)
    }

    /**
     * Write EntityProperties to output stream
     */
    async writeTo(stream: Stream) {
        let size = this.size;
        await stream.writeInt(size);
        let buildScore = this.buildScore;
        await stream.writeInt(buildScore);
        let destroyScore = this.destroyScore;
        await stream.writeInt(destroyScore);
        let canMove = this.canMove;
        await stream.writeBool(canMove);
        let populationProvide = this.populationProvide;
        await stream.writeInt(populationProvide);
        let populationUse = this.populationUse;
        await stream.writeInt(populationUse);
        let maxHealth = this.maxHealth;
        await stream.writeInt(maxHealth);
        let initialCost = this.initialCost;
        await stream.writeInt(initialCost);
        let sightRange = this.sightRange;
        await stream.writeInt(sightRange);
        let resourcePerHealth = this.resourcePerHealth;
        await stream.writeInt(resourcePerHealth);
        let build = this.build;
        if (build === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await build.writeTo(stream);
        }
        let attack = this.attack;
        if (attack === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await attack.writeTo(stream);
        }
        let repair = this.repair;
        if (repair === null) {
            await stream.writeBool(false);
        } else {
            await stream.writeBool(true);
            await repair.writeTo(stream);
        }
    }
}

File: generated-code/python/codecraft/TcpReadWrite/model/entity_properties.py
from model.attack_properties import AttackProperties
from model.build_properties import BuildProperties
from model.repair_properties import RepairProperties

class EntityProperties:
    """Entity properties"""

    def __init__(self, size, build_score, destroy_score, can_move, population_provide, population_use, max_health, initial_cost, sight_range, resource_per_health, build, attack, repair):
        self.size = size
        """Size. Entity has a form of a square with side of this length"""
        self.build_score = build_score
        """Score for building this entity"""
        self.destroy_score = destroy_score
        """Score for destroying this entity"""
        self.can_move = can_move
        """Whether this entity can move"""
        self.population_provide = population_provide
        """Number of population points this entity provides, if active"""
        self.population_use = population_use
        """Number of population points this entity uses"""
        self.max_health = max_health
        """Maximum health points"""
        self.initial_cost = initial_cost
        """Cost to build this first entity of this type. If this is a unit (entity can move), the cost is increased by 1 for each existing unit of this type"""
        self.sight_range = sight_range
        """If fog of war is enabled, maximum distance at which other entities are considered visible"""
        self.resource_per_health = resource_per_health
        """Amount of resource added to enemy able to collect resource on dealing damage for 1 health point"""
        self.build = build
        """Build properties, if entity can build"""
        self.attack = attack
        """Attack properties, if entity can attack"""
        self.repair = repair
        """Repair properties, if entity can repair"""

    @staticmethod
    def read_from(stream):
        """Read EntityProperties from input stream
        """
        size = stream.read_int()
        build_score = stream.read_int()
        destroy_score = stream.read_int()
        can_move = stream.read_bool()
        population_provide = stream.read_int()
        population_use = stream.read_int()
        max_health = stream.read_int()
        initial_cost = stream.read_int()
        sight_range = stream.read_int()
        resource_per_health = stream.read_int()
        if stream.read_bool():
            build = BuildProperties.read_from(stream)
        else:
            build = None
        if stream.read_bool():
            attack = AttackProperties.read_from(stream)
        else:
            attack = None
        if stream.read_bool():
            repair = RepairProperties.read_from(stream)
        else:
            repair = None
        return EntityProperties(size, build_score, destroy_score, can_move, population_provide, population_use, max_health, initial_cost, sight_range, resource_per_health, build, attack, repair)
    
    def write_to(self, stream):
        """Write EntityProperties to output stream
        """
        stream.write_int(self.size)
        stream.write_int(self.build_score)
        stream.write_int(self.destroy_score)
        stream.write_bool(self.can_move)
        stream.write_int(self.population_provide)
        stream.write_int(self.population_use)
        stream.write_int(self.max_health)
        stream.write_int(self.initial_cost)
        stream.write_int(self.sight_range)
        stream.write_int(self.resource_per_health)
        if self.build is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.build.write_to(stream)
        if self.attack is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.attack.write_to(stream)
        if self.repair is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.repair.write_to(stream)
    
    def __repr__(self):
        return "EntityProperties(" + \
            repr(self.size) + \
            ", " + \
            repr(self.build_score) + \
            ", " + \
            repr(self.destroy_score) + \
            ", " + \
            repr(self.can_move) + \
            ", " + \
            repr(self.population_provide) + \
            ", " + \
            repr(self.population_use) + \
            ", " + \
            repr(self.max_health) + \
            ", " + \
            repr(self.initial_cost) + \
            ", " + \
            repr(self.sight_range) + \
            ", " + \
            repr(self.resource_per_health) + \
            ", " + \
            repr(self.build) + \
            ", " + \
            repr(self.attack) + \
            ", " + \
            repr(self.repair) + \
            ")"